### PR TITLE
executor: make the memory tracker of Jsonobjectagg more accurate

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
We need estimate the memory usage of map[string]interface{} in partialResult4JsonObjectAgg, and trace it.

How it Works:

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- No release note

### CODE REVIEW

File: executor/aggfuncs/func_json_objectagg.go [hunk 1]
```diff
@@ -105,8 +110,10 @@ func (e *jsonObjectAgg) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup
 		switch x := realVal.(type) {
 		case nil, bool, int64, uint64, float64, string, json.BinaryJSON, *types.MyDecimal, []uint8, types.Time, types.Duration:
 			if _, ok := p.entries[keyString]; !ok {
-				memDelta += int64(len(keyString))
-				memDelta += getValMemDelta(realVal)
+				memDelta += int64(len(keyString)) + getValMemDelta(realVal)
+				if len(p.entries)+1 > (1<<p.bInMap)*loadFactorNum/loadFactorDen {
+					memDelta += (1 << p.bInMap) * DefMapStringInterfaceBucketSize
+				}
```
Comment (ichn-hu) [inline executor/aggfuncs/func_json_objectagg.go:116]: it seems you forget to increase p.bInMap?

File: executor/aggfuncs/func_json_objectagg.go [hunk 2]
```diff
@@ -150,6 +157,10 @@ func (e *jsonObjectAgg) MergePartialResult(sctx sessionctx.Context, src, dst Par
 	// and only the last value encountered is used with that key in the returned object
 	for k, v := range p1.entries {
 		p2.entries[k] = v
+		memDelta += int64(len(k)) + getValMemDelta(v)
+		if len(p2.entries)+1 > (1<<p2.bInMap)*loadFactorNum/loadFactorDen {
+			memDelta += (1 << p2.bInMap) * DefMapStringInterfaceBucketSize
+		}
```
Comment (ichn-hu) [inline executor/aggfuncs/func_json_objectagg.go:163]: ditto

**Review by ichn-hu — CHANGES_REQUESTED** [review]
Rest LGTM

**Review by ichn-hu — APPROVED** [review]

File: executor/aggfuncs/set_with_memory_usage.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aggfuncs
+
+const (
+	// Maximum average load of a bucket that triggers growth is 6.5.
+	// Represent as loadFactorNum/loadFactDen, to allow integer math.
+	loadFactorNum = 13
```
Comment (XuHuaiyu) [inline executor/aggfuncs/set_with_memory_usage.go:19]: Can we use the var defined in aggregate.go?

Comment (wshwsh12) [reply executor/aggfuncs/set_with_memory_usage.go]: No. It will cause cycle import.
I try to move the two var to package `set` and export it. PTAL.

Comment (XuHuaiyu) [reply executor/aggfuncs/set_with_memory_usage.go]: Maybe we can move it to util/hack.go

And add a comment for these 2 consts to describe that they are from the golang definition.

**Review by XuHuaiyu — APPROVED** [review]
LGTM

**Pull request merged**